### PR TITLE
Feature/#19 add pypi package

### PR DESCRIPTION
**Pull request by prozessorkern**
addressing #19 

- Added setup.py script to create a pypi packet.
- Restructured the whole project and import structure
- removed the test plugin from the pypi packet
- left the entrypoint metrix++.py to be still able to run the tests
- added github actions to run the tests and send and send the package to pypi
- added action batches and pypi documentation to the readme.md

Please take a look at the changes if this breaks anything which I have overlooked.
Please check if the pypi packet works on your system (on linux it works fine).
[pypi packet](https://pypi.org/project/metrixpp/)

I uploaded the packet using my account. We should discuss how to handle the credentials to pypi.

### CODE REVIEW

File: .github/workflows/python-package.yml [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    steps:
+    - uses: actions/checkout@v2
+    - name: Set up Python
+      uses: actions/setup-python@v2
+      with:
+        python-version: '3.x'
+    - name: Install dependencies
+      run: |
+        python -m pip install --upgrade pip
+        pip install setuptools wheel twine
```
Comment (avkonst) [inline .github/workflows/python-package.yml:24]: what is wheel and twine?

File: .github/workflows/python-tests.yml [hunk 2]
```diff
@@ -0,0 +1,32 @@
+# This workflow will install Python dependencies, run tests and lint with a single version of Python
```
Comment (avkonst) [inline .github/workflows/python-tests.yml:1]: are these workflows free of charge?

File: .github/workflows/python-tests.yml [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+# For more information see: https://help.github.com/actions/language-and-framework-guides/using-python-with-github-actions
```
Comment (avkonst) [inline .github/workflows/python-tests.yml:2]: is github showing a status if the workflow action failed / passed?

File: README.md [hunk 4]
```diff
@@ -1,12 +1,23 @@
 # [Metrix++](https://github.com/metrixplusplus/metrixplusplus)
+
+![Metrix++ Tests](https://github.com/metrixplusplus/metrixplusplus/workflows/Metrix++%20Tests/badge.svg)
```
Comment (avkonst) [inline README.md:3]: very nice!

File: setup.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
+import setuptools
+
+with open("README.md", "r") as fh:
+    long_description = fh.read()
+
+setuptools.setup(
+    name="metrixpp",
+    version="0.0.1a",
+    author="Stefan Strobel",
```
Comment (avkonst) [inline setup.py:9]: is it your name? just checking that you have not left someone else's name from an example :)

File: setup.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
+import setuptools
+
+with open("README.md", "r") as fh:
+    long_description = fh.read()
+
+setuptools.setup(
+    name="metrixpp",
+    version="0.0.1a",
```
Comment (avkonst) [inline setup.py:8]: When ready, it is better to continue from the version we had.. Feel free to make a release via the next github tag..

File: setup.py [hunk 7]
```diff
@@ -0,0 +1,38 @@
+import setuptools
```
Comment (avkonst) [inline setup.py:1]: have you verified it yourself on clean install?

File: setup.py [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+    name="metrixpp",
+    version="0.0.1a",
+    author="Stefan Strobel",
+    author_email="stefan.strobel@shimatta.net",
+    description="Metrix++ is an extendable tool for code metrics collection and analysis.",
+    long_description=long_description,
+    long_description_content_type="text/markdown",
+    url="https://github.com/metrixplusplus/metrixplusplus",
+    packages=setuptools.find_packages(exclude=["metrixpp.tests"]),
+    package_data={
```
Comment (avkonst) [inline setup.py:16]: is somebody adds a new folder (unlikely but still), how do we verify that the bundled package is still functioning?

File: setup.py [hunk 9]
```diff
@@ -0,0 +1,38 @@
+import setuptools
+
+with open("README.md", "r") as fh:
+    long_description = fh.read()
+
+setuptools.setup(
```
Comment (avkonst) [inline setup.py:6]: Regarding pypi credentials. I hope you will continue to look after the project as you have got more understanding and visibility to what is going on with the metrix++ tool and python world. If there is a way to share an access with me to the  python packages repository, please share it with me just for safety / security / back up purposes. (Let me know what I should do, so you can share it with me). I am not intending to publish metrixpp package myself, but who knows when and if you need help... It is good that you have automated it (as far as I understand).

File: setup.py [hunk 10]
```diff
@@ -0,0 +1,38 @@
+import setuptools
+
```
Comment (avkonst) [inline setup.py:2]: I will also ask somebody else to verify. Will keep you in CC

**Comment by avkonst** [comment]
Hi @clegaard please verify that metrix++ works from the pypi package. @prozessorkern has completed what you attempted to do back 2-3 months ago. Please help us to verify that it works as you expected it from the pypi package.

File: .github/workflows/python-package.yml (hunk 1 above)
Comment (prozessorkern) [reply .github/workflows/python-package.yml:24]: wheel: needed for python packaging
twine: tool to upload the packet on pypi

File: .github/workflows/python-tests.yml (hunk 2 above)
Comment (prozessorkern) [reply .github/workflows/python-tests.yml:1]: For public repositories they are free:
https://help.github.com/en/github/setting-up-and-managing-billing-and-payments-on-github/about-billing-for-github-actions

File: .github/workflows/python-tests.yml (hunk 3 above)
Comment (prozessorkern) [reply .github/workflows/python-tests.yml:2]: I added badges to the readme file so the current status is shown in github and pypi (they are showing the readme file as documentation)

File: setup.py (hunk 5 above)
Comment (prozessorkern) [reply setup.py:9]: Yes, that's me ;)

File: setup.py (hunk 6 above)
Comment (prozessorkern) [reply setup.py:8]: As soon as this is on master I will create a release and set the packet version to the new release version (I would suggest 1.6.0)

File: setup.py (hunk 7 above)
Comment (prozessorkern) [reply setup.py:1]: not on my system, but the github action runs from a clean system and it seems to work fine.

File: setup.py (hunk 8 above)
Comment (prozessorkern) [reply setup.py:16]: this is a good point.
I did not include the tests into the packet to reduce the size (and the config is quite bad if you want to include non *.py files)

It is though possible to run the tests from the repo against the package to verify nothing is missing...
I will try to add this to the action pipeline to create a local packet on every build and test against it.

File: setup.py (hunk 9 above)
Comment (prozessorkern) [reply setup.py:6]: Just create an account on pypi and tell me your username. I will then share the project with you.
https://pypi.org/account/register/

**Comment by avkonst** [comment]
I have created avkonst user in pypi.

**Comment by avkonst** [comment]
I understand you have got permissions to merge it, right? Feel free to merge when you think you are ready. Thanks for this great feature.

**Pull request merged**